### ISSUE TITLE
Unchecking Credit card / debit card in Payments accepted on checkout does not remove it at checkout 

### ISSUE THREAD

**Issue by darcie** [Automattic/woocommerce-payments#3969]
### Describe the bug
When Credit card / debit card in Payments accepted on checkout is unchecked and Express checkouts is enabled, the option to pay with a card still appears to the customer. 

### To Reproduce

1. Go to Payments > Settings
2. Uncheck the option to display credit/debit cards. 
3. You'll see a warning message that customers will no longer be able to use this method for payment. 
4. Visit checkout
5. The credit card fields are still visible and can be used for payment. 

Tested with and without UPE. 

### Actual behavior

Some merchants only want to use WooCommerce Payments for Express Checkout features. While Credit card / debit card can be deselected, this doesn't seem to have any impact on availability at checkout. 

### Screenshots

Screenshot of warning when the payment method is removed: 
<img width="833" alt="Screenshot on 2022-03-15 at 15-37-20" src="https://user-images.githubusercontent.com/318285/158485022-2fb18ab9-c9d0-413c-887b-739baca5eeae.png">

Screenshot of the checkout in Chrome, Express is available but credit card is also. 
<img width="870" alt="Markup 2022-03-15 at 15 39 01" src="https://user-images.githubusercontent.com/318285/158484952-d4f13976-8505-49b0-bd60-26acee03123d.png">

### Expected behavior

Since credit/debit has been disabled, we expect that it would not appear. 
If it cannot be fully disabled, I would expect the checkbox to not appear for that payment method. 

### Desktop (please complete the following information):

* OS: [e.g. iOS] Mac OS 12.2.1
* Browser [e.g. chrome, safari] Chrome 
* Version [e.g. 22] Version 99.0.4844.51

### Additional context

Brought up in 4862961-zen 
Previously discussed in https://github.com/Automattic/woocommerce-payments/issues/2560#issuecomment-896222148
And: https://wordpress.org/support/topic/only-apple-google-pay-on-woocommerce/

**Comment by aheckler** [comment]
4868656-zen

**Comment by mcampal** [comment]
Also reported here: 4893444-zen

**Comment by jessy-p** [comment]
Marking for Product Review to confirm: https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1072568063

**Comment by dreamtooloud** [comment]
Reported here: hc-35065471

**Comment by yukikatayama** [comment]
@LevinMedia - do we have an update regarding [this comment](https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1072568063)?

**Comment by AashikP** [comment]
Just noting one use case: Where merchants only want to enable Express checkout buttons on their site Google/Apple Pay.

**Comment by rogermattic** [comment]
Only enabling e.g. Google Pay, but not a card payment IMHO is a good way to not get paid BUT as weird as this seems to me now, everything can happen and since we have this flow confirmation, that informs the user that they can disable a card payment... we should probably support that use case in the checkout - so not display the card inputs there, when the user intentionally disabled them.

**Comment by sverleis** [comment]
Another user with this issue: 5203578-zen

Offered them some CSS to hide the option: 

```
.payment_method_woocommerce_payments {
    display: none;
}
```
This user is specifically using PayPal for their card transactions.

This issue can also be replicated on the Stripe Extension:

```
.payment_method_stripe {
	display: none
}
```

**Comment by RadoslavGeorgiev** [comment]
@jessy-p could we handle this as a part of the UPE project?

**Comment by jessy-p** [comment]
paJDYF-4kY-p2

**Comment by apmwebdev** [comment]
Another report in 29399875-hc/5321338-zen.

**Comment by kgeorgee** [comment]
Reported here: 5331014-zen

**Comment by prnvrna** [comment]
Issue reported here as well: 5482575-zen.

**Comment by csmcneill** [comment]
In 5556446-zen, with UPE enabled and `Credit card / debit card` _disabled_, this produced tons of console errors:

```
upe_checkout.js?ver=4.7.2:1 Uncaught TypeError: Cannot read properties of undefined (reading 'title')
    at upe_checkout.js?ver=4.7.2:1:107599
    at I (upe_checkout.js?ver=4.7.2:1:107624)
    at HTMLBodyElement.<anonymous> (upe_checkout.js?ver=4.7.2:1:107800)
    at HTMLBodyElement.dispatch (jquery.min.js:2:43064)
    at v.handle (jquery.min.js:2:41048)
    at Object.trigger (jquery.min.js:2:71515)
    at HTMLBodyElement.<anonymous> (jquery.min.js:2:72110)
    at Function.each (jquery.min.js:2:3003)
    at s.fn.init.each (jquery.min.js:2:1481)
    at s.fn.init.trigger (jquery.min.js:2:72086)
```

```
?ver=3.0:1 Uncaught TypeError: Cannot read properties of undefined (reading 'isReusable')
    at upe_checkout.js?ver=4.7.2:1:107296
    at ?ver=3.0:1:135460
    at Array.forEach (<anonymous>)
    at n._emit (?ver=3.0:1:135401)
    at i._emitEvent (?ver=3.0:1:245767)
    at ?ver=3.0:1:232339
    at ?ver=3.0:1:135658
    at Array.forEach (<anonymous>)
    at n._emit (?ver=3.0:1:135401)
    at e._handleMessage (?ver=3.0:1:265503)
```

This also prevented any payments from being accepted on checkout. Trying to force a `402` error by inputting test card information while WCPay isn't in test mode yields an additional console error:

```
?ver=3.0:1 Uncaught (in promise) IntegrationError: You specified "never" for fields.billing_details.name when creating the payment Element, but did not pass confirmParams.payment_method_data.billing_details.name when calling stripe.confirmPayment(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.confirmPayment().
    at $r (?ver=3.0:1:261803)
    at e._handleMessage (?ver=3.0:1:266688)
    at e._handleMessage (?ver=3.0:1:133942)
    at ?ver=3.0:1:264640
```

In this instance, CSS to hide the payment option may not work if the user is trying to use UPE to accept other payment methods (e.g. Sofort).

**Comment by NathalyaVides** [comment]
Reported here 5658637-zen

**Comment by zoemyers** [comment]
Discovered this was the problem in 37682003-hc

**Comment by ericfri** [comment]
Also in 5723093-zen

**Comment by amElnagdy** [comment]
And in 5754784-zen

**Comment by devvnawaz** [comment]
Reported in 5788340-zen

**Comment by zoemyers** [comment]
Another one in 5805377-zen

**Comment by mouligreenlaw** [comment]
5912855-zen

**Comment by NathalyaVides** [comment]
5959477-zen

**Comment by naman03malhotra** [comment]
Thanks for [extensive research and calculating the estimates](https://github.com/Automattic/woocommerce-payments-meta/issues/68#issuecomment-1445733734) on this issue @jessy-p. Looks like a significant effort will go into enabling this use case. Given the [work around is available](https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1128129498). I will defer to @souravdebnath1986 to help align the priority from a product perspective.

**Comment by naman03malhotra** [comment]
cc Checking in with @Automattic/fusion, does this fall under your umbrella and be handled as part of your backlog?

**Comment by anu-rock** [comment]
@naman03malhotra Thanks for your ping. While we wait for @souravdebnath1986's inputs on prioritization, am I correct in understanding that the issue isn't directly related to Apple/Google Pay, but we'll benefit in fixing for them as they are the only non-UPE payment methods for now?

**Comment by Octolus** [comment]
I have this issue, i have a credit card enabled in Stripe also. However, when trying to use Google Pay or Apple Pay, it throws this error and does not redirect/complete the checkout.

```caught (in promise) IntegrationError: You specified "never" for fields.billing_details.name when creating the payment Element, but did not pass confirmParams.payment_method_data.billing_details.name when calling stripe.confirmPayment(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.confirmPayment().```

Are there any workarounds for my case? I tried to disable all plugins and whatever, but it's not working.

**Comment by naman03malhotra** [comment]
As discussed with @souravdebnath1986 during the call, he'll get back on the priority of this issue from a product perspective.

**Comment by souravdebnath1986** [comment]
This looks like an edge case with a workaround available for hiding the card CSS. Given the effort to fix is high, I would prioritise when we have sprint capacity available that doesn't impact project work.

**Comment by naman03malhotra** [comment]
Thanks, @souravdebnath1986, I am putting this back to "prioritization queue", tbd in the sync call with @Automattic/gamma

**Comment by xue28** [comment]
6077903-zen

**Comment by darcie** [comment]
Request to only use express buttons in 6266117-zen

**Comment by francoishvz** [comment]
Merchant only wants to use express checkout options: 6441117-zd-woothemes

**Comment by maxlaf** [comment]
6452109-zen

**Comment by naman03malhotra** [comment]
@csmcneill I am trying to assess the priority of this issue since we have incoming tickets. Given the high effort to solve this and work around being available, does priority remains the same as we [discussed before](https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1463441141)?

**Comment by csmcneill** [comment]
@naman03malhotra I would say so, especially since there are [available workarounds](https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1128129498).

**Comment by Babylon1999** [comment]
one more in zd-6495287

**Comment by darcie** [comment]
Request in 6807801-zen

**Comment by jtobiesen** [comment]
Requested in 7315964-zd-a8c

**Comment by csmcneill** [comment]
@naman03malhotra Given that...

1. UPE is now standard
2. The card payment option can no longer be disabled

...I'm not sure that this issue should remain open unless we (at some point) want to allow merchants the ability to disable the card payment option and only allow payments Google/Apple Pay.

**Comment by htdat** [comment]
Thanks @csmcneill - I am moving this issue back to queue "Ready for prioritization" so that Gamma can re-evaluate it. 

> UPE is now standard

To avoid confusion as we have more than one UPE in terms of technical architecture, this is exactly `Split-UPE-with-Deferred-Intent`. 
Ref: "Glossary & visual reference of supported checkout flavors" of this post paJDYF-9xf-p2

**Comment by frosso** [comment]
Something to consider - when the CC is disabled, Stripe Link might behave inconsistently, since it relies on the same gateway classes.

**Comment by Prin-S** [comment]
7349724-zd-a8c

**Comment by naman03malhotra** [comment]
Workaround is available here: https://github.com/Automattic/woocommerce-payments/issues/3969#issuecomment-1128129498

Ref: paJDYF-buB-p2#comment-22004

**Comment by pierorocca** [comment]
> Something to consider - when the CC is disabled, Stripe Link might behave inconsistently, since it relies on the same gateway classes.

Noting that hiding rather than disabling would be preferred in order to allow WooPay to be functional. Stripe Link may be mutually exclusive however since it's expecting the CC fields in the checkout form whereas WooPay does not.

**Comment by timur27** [comment]
I was passing by and wanted to share some thoughts. I think the current state is that it's impossible to disable credit cards. However, we can assume that our system should be prepared for such a scenario. In this case, as of my knowledge and tests today, the below behavior is to be observed when CC is _hypothetically_ disabled: 

<p align="center">
<img src="https://github.com/Automattic/woocommerce-payments/assets/22319444/ddc941f7-9df9-429e-bd0e-6711b3600ea8"  height="400" />
</p>


The above behavior happens because WooCommerce Core doesn't know if CC is active or not, it assumes it's active even after we disable CC from settings. I guess we discussed the need to improve this in the thread to handle refunds for disabled PMs (paJDYF-bPM-p2). I described this problem in more detail in paJDYF-bPM-p2#comment-22640. It is possible that the situation discussed in this issue will naturally improve once refunds for disabled PMs are implemented. Please consider this as a potential solution, as I'm not 100% sure what the final implementation will look like. 


To sum up, I think one of the options with this issue is
* to wait until refunding for disabled payment methods (paJDYF-bPM-p2) is implemented
* simulate disabling CC from the settings level by modifying the settings POST request's body
* confirm that even with this potential scenario with CC being disabled, the gateway is not rendered on the checkout page

> Noting that hiding rather than disabling would be preferred in order to allow WooPay to be functional

The way settings update will stop showing the payment method on the checkout after implementing the feature for refunds paJDYF-bPM-p2 will allow WooPayments to treat the card as if it was enabled, which will allow us to pass the scenario outlined by Piero.

**Comment by pierorocca** [comment]
@timur27 thanks for the additional time thinking through this use case. Is there a way to only hide, not disable the Credit/Debit, options? It's only shoppers that we'd want to suppress credit/debit from being directly accessed while allowing indirect access through PRBs, Apple Pay/Google Pay, and full back-end capability to manage disputes, refunds, payouts, etc.

**Comment by timur27** [comment]
@pierorocca Yes, I think this is what the solution for the refunds P2 I linked above might eventually allow us to do. Checkboxes in Settings screen will only hide/show PMs on checkouts, while PMs will be available for the plugin during the runtime. This will allow us to implement credit card toggling as well in the next step (if this feature will be needed).

We discussed the approach of only hiding PMs high-level so the scope/outcome might change depending on the final implementation Team Pulsar will proceed with.

**Comment by Thelmachido** [comment]
Another request 7777831-zen

**Comment by rogermattic** [comment]
Pinging @nikkivias here as she's currently working on some checkout improvements :)

**Comment by dsmithweb-a8c** [comment]
8624859-zen

**Comment by AashikP** [comment]
Another request in the form of a forum review [here](https://wordpress.org/support/topic/jail-plugin-avoid/).

**Comment by pierorocca** [comment]
@FangedParakeet let's get this one in Sprint 15 if possible. Settings are already geared for this use case so it should be about how to hide the component on the front-end and Blocks editor.

**Comment by FangedParakeet** [comment]
@pierorocca, the description of this issue is somewhat out of date. From my testing, it is no longer possible to disable the CC payment gateway from payments settings. 

I am aware that this feature is a very highly requested enhancement, so I do not want to let it be forgotten. That being said, while this feature is trivial in V2, in V1 this would necessitate a substantial refactor to separate the CC payment method into its own gateway separate from plugin functionality. This would not only require a new gateway, but also the ability to disambiguate the plugin from the CC method and to find a new home for admin and checkout hooks for all PMs as well as checkout classes that enqueue scripts and assets.

Since this effort would likely require a project and epic of its own to complete, I would like to discuss the viability of implementing this in V2--where this will be much more trivial, as currently designed--against the much greater investment to implement this in the current plugin. Once this is prioritised, we can begin scoping out this effort assigning to a future sprint. I am keeping this in our backlog for the time being until we have more clarity on this decision.

**Comment by pierorocca** [comment]
IIRC the idea here was to hide the card component from users but keep it enabled thus still enabling express checkouts. Is that a much smaller lift? Any drawbacks?

**Comment by FangedParakeet** [comment]
> IIRC the idea here was to hide the card component from users but keep it enabled thus still enabling express checkouts.

Ah, cool. This could be an available short-term fix to simply hide the gateway at checkout, though this would merely be a cosmetic change. The problem is that we are basically still forced to register the CC gateway in order to maintain functionality of the plugin. With other APMs, they are _enabled_, so that refunds, etc. can be processed, but they are no longer registered as payment gateways available for checkout. 

We _could_ hide the gateway at checkout, but this would only be hiding the elements on the DOM. WC core would still load the CC payment gateway onto the DOM, since the gateway remains registered, and all CC API endpoints to process payments would remain available. This basically means that the CC UI would be hidden from the checkout page, but users would not necessarily be prevented from completing transactions with the CC PM via manipulating the DOM or using the Store API.

Ideally, the longer term fix would be not to register the CC gateway at all, as we do with other PMs, so that it would be truly disabled, rather than simply making the elements invisible to the UI. 

Let me know if that makes sense and sounds worthwhile to achieve, @pierorocca. 🙏

**Comment by pierorocca** [comment]
Agree with both of your long views. Thanks for contemplating the cons.

In the short term, if a user was savvy enough to manipulate checkout at the DOM level or through the API to pay by card, that's totally fine imo. My understanding is that the merchant is trying to manage the shopper experience and not expose duplicate payment methods while STILL accepting cards through express payment methods. It's a workaround for being vendor locked.

IF we did pursue this through short term means, if don't think using the disable checkbox is the right approach since it is in fact enabled, just not visible to shoppers in the UI.

In terms of the issue reported in this ticket, I think we can close it since it's solved or we convert it to address the real ask in the Zendesk ticket?

**Comment by FangedParakeet** [comment]
> IF we did pursue this through short term means, if don't think using the disable checkbox is the right approach since it is in fact enabled, just not visible to shoppers in the UI.

Agreed here. Suggesting that the CC gateway is disabled when it is merely hidden cosmetically could have unintended effects, particularly if store owners might incur particular fees for CC payments that they might be trying to avoid by deselecting the gateway.

> In terms of the issue reported in this ticket, I think we can close it since it's solved or we convert it to address the real ask in the Zendesk ticket?

Also, agreed here. I think the original symptoms diagnosed in this issue are no longer relevant nor present. We can continue discussing the available approaches towards resolving the tangential purpose of disabling CC in favour of EPMs on Slack (see p1730972949522029-slack-C043JN7AJHY) and if we agree that this approach is indeed agreeable and viable, I'll create a new issue for it. 👍 

TL;DR: Closing this issue as resolved and no longer relevant.
